### DISCUSSION TITLE
Ollama compatible?

### DISCUSSION THREAD

**Discussion by Masber** [yetone/avante.nvim#687] [Q&A] [answered]
Hi,

I am looking for a AI assistant for my neovim setup. I was wondering whether I could use avant.nvim with local LLM models through llama?

thank you

**Comment by ha100** [comment, marked as answer]
sure. fully functional lazy config looks like this

```lua
return {
    "yetone/avante.nvim",
    event = "VeryLazy",
    lazy = false,
    version = false, -- set this if you want to always pull the latest change
    opts = {
        provider = "ollama",
        use_absolute_path = true,
        vendors = {
            ---@type AvanteProvider
            ollama = {
                ['local'] = true,
                endpoint = "http://localhost:11434/v1",
                model = "codellama:7b-instruct",
                parse_curl_args = function(opts, code_opts)
                    return {
                        url = opts.endpoint .. "/chat/completions",
                        headers = {
                            ["Accept"] = "application/json",
                            ["Content-Type"] = "application/json",
                            ['x-api-key'] = 'ollama',
                        },
                        body = {
                            model = opts.model,
                            messages = require("avante.providers").copilot.parse_message(code_opts), -- you can make your own message, but this is very advanced
                            max_tokens = 2048,
                            stream = true,
                        },
                    }
                end,
                parse_response_data = function(data_stream, event_state, opts)
                    require("avante.providers").openai.parse_response(data_stream, event_state, opts)
                end,
            },
        },
        behaviour = {
            auto_suggestions = false, -- Experimental stage
            auto_set_highlight_group = true,
            auto_set_keymaps = true,
            auto_apply_diff_after_generation = false,
            support_paste_from_clipboard = true,
        },
        mappings = {
            --- @class AvanteConflictMappings
            diff = {
                ours = 'co',
                theirs = 'ct',
                all_theirs = 'ca',
                both = 'cb',
                cursor = 'cc',
                next = ']x',
                prev = '[x',
            },
            suggestion = {
                accept = '<M-l>',
                next = '<M-]>',
                prev = '<M-[>',
                dismiss = '<C-]>',
            },
            jump = {
                next = ']]',
                prev = '[[',
            },
            submit = {
                normal = '<CR>',
                insert = '<C-s>',
            },
        },
        hints = { enabled = true },
        windows = {
            ---@type "right" | "left" | "top" | "bottom"
            position = 'right',   -- the position of the sidebar
            wrap = true,          -- similar to vim.o.wrap
            width = 30,           -- default % based on available width
            sidebar_header = {
                align = 'center', -- left, center, right for title
                rounded = true,
            },
        },
        highlights = {
            ---@type AvanteConflictHighlights
            diff = {
                current = 'DiffText',
                incoming = 'DiffAdd',
            },
        },
        --- @class AvanteConflictUserConfig
        diff = {
            autojump = true,
            ---@type string | fun(): any
            list_opener = 'copen',
        },
    },
    -- if you want to build from source then do `make BUILD_FROM_SOURCE=true`
    build = "make BUILD_FROM_SOURCE=true",
    -- build = "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false" -- for windows
    dependencies = {
        "nvim-treesitter/nvim-treesitter",
        "stevearc/dressing.nvim",
        "nvim-lua/plenary.nvim",
        "MunifTanjim/nui.nvim",
        --- The below dependencies are optional,
        "nvim-tree/nvim-web-devicons", -- or echasnovski/mini.icons
        {
            -- support for image pasting
            "HakonHarnes/img-clip.nvim",
            event = "VeryLazy",
            opts = {
                -- recommended settings
                default = {
                    embed_image_as_base64 = false,
                    prompt_for_file_name = false,
                    drag_and_drop = {
                        insert_mode = true,
                    },
                    -- required for Windows users
                    use_absolute_path = true,
                },
            },
        },
        {
            -- Make sure to set this up properly if you have lazy=true
            'MeanderingProgrammer/render-markdown.nvim',
            opts = {
                file_types = { "markdown", "Avante" },
            },
            ft = { "markdown", "Avante" },
        },
    },
    keys = {
        {
            "<leader>aa",
            function()
                require("avante.api").ask()
            end,
            desc = "avante: ask",
            mode = { "n", "v" },
        },
        {
            "<leader>ar",
            function()
                require("avante.api").refresh()
            end,
            desc = "avante: refresh",
        },
        {
            "<leader>ae",
            function()
                require("avante.api").edit()
            end,
            desc = "avante: edit",
            mode = "v",
        },
    },
}

```

**Reply by Marvinus** [reply]
> sure. fully functional lazy config looks like this
> 
> ```lua
> return {
>     "yetone/avante.nvim",
>     event = "VeryLazy",
>     lazy = false,
>     version = false, -- set this if you want to always pull the latest change
>     opts = {
>         provider = "ollama",
>         use_absolute_path = true,
>         vendors = {
>             ---@type AvanteProvider
>             ollama = {
>                 ['local'] = true,
>                 endpoint = "http://localhost:11434/v1",
>                 model = "codellama:7b-instruct",
>                 parse_curl_args = function(opts, code_opts)
>                     return {
>                         url = opts.endpoint .. "/chat/completions",
>                         headers = {
>                             ["Accept"] = "application/json",
>                             ["Content-Type"] = "application/json",
>                             ['x-api-key'] = 'ollama',
>                         },
>                         body = {
>                             model = opts.model,
>                             messages = require("avante.providers").copilot.parse_message(code_opts), -- you can make your own message, but this is very advanced
>                             max_tokens = 2048,
>                             stream = true,
>                         },
>                     }
>                 end,
>                 parse_response_data = function(data_stream, event_state, opts)
>                     require("avante.providers").openai.parse_response(data_stream, event_state, opts)
>                 end,
>             },
>         },
>         behaviour = {
>             auto_suggestions = false, -- Experimental stage
>             auto_set_highlight_group = true,
>             auto_set_keymaps = true,
>             auto_apply_diff_after_generation = false,
>             support_paste_from_clipboard = true,
>         },
>         mappings = {
>             --- @class AvanteConflictMappings
>             diff = {
>                 ours = 'co',
>                 theirs = 'ct',
>                 all_theirs = 'ca',
>                 both = 'cb',
>                 cursor = 'cc',
>                 next = ']x',
>                 prev = '[x',
>             },
>             suggestion = {
>                 accept = '<M-l>',
>                 next = '<M-]>',
>                 prev = '<M-[>',
>                 dismiss = '<C-]>',
>             },
>             jump = {
>                 next = ']]',
>                 prev = '[[',
>             },
>             submit = {
>                 normal = '<CR>',
>                 insert = '<C-s>',
>             },
>         },
>         hints = { enabled = true },
>         windows = {
>             ---@type "right" | "left" | "top" | "bottom"
>             position = 'right',   -- the position of the sidebar
>             wrap = true,          -- similar to vim.o.wrap
>             width = 30,           -- default % based on available width
>             sidebar_header = {
>                 align = 'center', -- left, center, right for title
>                 rounded = true,
>             },
>         },
>         highlights = {
>             ---@type AvanteConflictHighlights
>             diff = {
>                 current = 'DiffText',
>                 incoming = 'DiffAdd',
>             },
>         },
>         --- @class AvanteConflictUserConfig
>         diff = {
>             autojump = true,
>             ---@type string | fun(): any
>             list_opener = 'copen',
>         },
>     },
>     -- if you want to build from source then do `make BUILD_FROM_SOURCE=true`
>     build = "make BUILD_FROM_SOURCE=true",
>     -- build = "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false" -- for windows
>     dependencies = {
>         "nvim-treesitter/nvim-treesitter",
>         "stevearc/dressing.nvim",
>         "nvim-lua/plenary.nvim",
>         "MunifTanjim/nui.nvim",
>         --- The below dependencies are optional,
>         "nvim-tree/nvim-web-devicons", -- or echasnovski/mini.icons
>         {
>             -- support for image pasting
>             "HakonHarnes/img-clip.nvim",
>             event = "VeryLazy",
>             opts = {
>                 -- recommended settings
>                 default = {
>                     embed_image_as_base64 = false,
>                     prompt_for_file_name = false,
>                     drag_and_drop = {
>                         insert_mode = true,
>                     },
>                     -- required for Windows users
>                     use_absolute_path = true,
>                 },
>             },
>         },
>         {
>             -- Make sure to set this up properly if you have lazy=true
>             'MeanderingProgrammer/render-markdown.nvim',
>             opts = {
>                 file_types = { "markdown", "Avante" },
>             },
>             ft = { "markdown", "Avante" },
>         },
>     },
>     keys = {
>         {
>             "<leader>aa",
>             function()
>                 require("avante.api").ask()
>             end,
>             desc = "avante: ask",
>             mode = { "n", "v" },
>         },
>         {
>             "<leader>ar",
>             function()
>                 require("avante.api").refresh()
>             end,
>             desc = "avante: refresh",
>         },
>         {
>             "<leader>ae",
>             function()
>                 require("avante.api").edit()
>             end,
>             desc = "avante: edit",
>             mode = "v",
>         },
>     },
> }
> ```

I am guessing the same process applies to LM Studio ?

**Reply by SerkanSipahi** [reply]
@ha100 @Marvinus did it work for you? Is it good?

I am asking because the readme is saying: 

https://github.com/yetone/avante.nvim?tab=readme-ov-file#usage

> Avante will only support Claude, and OpenAI (and its variants including azure)out-of-the-box due to its high code quality generation. For all OpenAI-compatible providers, see [wiki](https://github.com/yetone/avante.nvim/wiki) for more details.

> Due to the poor performance of other models, avante.nvim only recommends using the claude-3.5-sonnet model. All features can only be guaranteed to work properly on the claude-3.5-sonnet model. We do not accept changes to the code or prompts to accommodate other models.

**Reply by Alex23rodriguez** [reply]
> I am guessing the same process applies to LM Studio ?

for anyone who landed here looking to setup LM Studio, this currently works (v0.0.19)

```lua
{
  -- in lazy settings for avante
  opts = {
    provider = 'lmstudio',
    vendors = {
      ---@type AvanteProvider
      lmstudio = {
        __inherited_from = 'openai',
        api_key_name = '',
        endpoint = 'http://127.0.0.1:1234/v1',
        model = '<model name here>',
      },
    },
  },
}
```

**Comment by ProgmRuanSilva** [comment]
# Error on `parse_message`
Hi everyone!! 🙃 

So, I had a problem with `parse_message`

After looking through the entire repository, I was able to find that the error is in the name of the function, which is now `parse_messages`

Another thing for anyone who is having this error now is the copilot dependency that is required

**Comment by abair** [comment]
I had the same problem, and this post got me past it to a new problem. With the right function name and copilot added as a dep, I get the following error when calling `:AvanteAsk` and sending it a query. I'm trying to set up a custom provider to use Ollama local on my machine:

E5108: Lua: ...ack/core/opt/avante.nvim/lua/avante/providers/openai.lua:192: attempt to index local 'opts' (a nil value)

Here's the config:

```
vim.pack.add({
    -- Required dependencies
    'https://github.com/nvim-lua/plenary.nvim',
    'https://github.com/MunifTanjim/nui.nvim',
    'https://github.com/zbirenbaum/copilot.lua',
})
require('copilot').setup({})

vim.pack.add({'https://github.com/yetone/avante.nvim'})
require('avante').setup({
    -- this file can contain specific instructions for your project
    instructions_file = 'avante.md',
    provider = 'ollama_local',
    -- provider = 'ollama',
    providers = {
        -- ollama = {
        --     ['local'] = true,
        --     is_env_set = require('avante.providers.ollama').check_endpoint_alive,
        --     endpoint = 'http://localhost:11434',
        --     model = 'qwen3.6:latest',
        --     timeout = 30000, -- Timeout in milliseconds
        --     extra_request_body = {
        --         temperature = 0.75,
        --         -- max_tokens = 20480,
        --     },
        -- },
        ollama_local = {
            ['local'] = true,
            endpoint = 'http://localhost:11434',
            model = 'qwen3.6',
            is_env_set = require("avante.providers.ollama").check_endpoint_alive,
            timeout = 30000, -- Timeout in milliseconds
                extra_request_body = {
                    temperature = 0.75,
                    -- max_tokens = 20480,
                },
            parse_curl_args = function(opts, code_opts)
                return {
                    url = opts.endpoint .. "/chat/completions",
                    headers = {
                        ["Accept"] = "application/json",
                        ["Content-Type"] = "application/json",
                        ['x-api-key'] = 'ollama',
                    },
                    body = {
                        model = opts.model,
                        messages = require("avante.providers").copilot.parse_messages(code_opts), -- you can make your own message, but this is very advanced
                        -- max_tokens = 2048,
                        stream = true,
                    },
                }
            end,
            parse_response_data = function(data_stream, event_state, opts)
                require("avante.providers").openai.parse_response(data_stream, event_state, opts)
            end,
        },
    },
})
```
